packages/flutter_tools/lib/src/base/error_handling_io.dart

```diff
@@ -671,3 +755,91 @@ void _throwFileSystemException(String errorMessage) {
   }
   throwToolExit(errorMessage);
 }
+
+
+/// Searches the `PATH` for the executable that [command] is supposed to launch.
+///
+/// This first builds a list of candidate paths where the executable may reside.
+/// If [command] is already an absolute path, then the `PATH` environment
+/// variable will not be consulted, and the specified absolute path will be the
+/// only candidate that is considered.
+///
+/// Once the list of candidate paths has been constructed, this will pick the
+/// first such path that represents an existent file.
+///
+/// If [strict] is false, return The original executable if there were no viable
+/// candidates - otherwise return `null`
+@visibleForTesting
+String resolveExecutablePath(
+  String command,
+  String workingDirectory, {
+  @required FileSystem fileSystem,
+  @required Platform platform,
+  @required Logger logger,
+  @required bool strict,
+}) {
+  try {
+    ErrorHandlingFileSystem.noExitOnFailure(() {
+      workingDirectory ??= fileSystem.currentDirectory.path;
+    });
+  } on Exception {
+    // The `currentDirectory` getter can throw a FileSystemException for example
+    // when the process doesn't have read/list permissions in each component of
+    // the cwd path. In this case, fall back on '.'.
+    workingDirectory ??= '.';
+  }
+  final String pathSeparator = platform.isWindows ? ';' : ':';
+
+  List<String> extensions = <String>[];
+  if (platform.isWindows && fileSystem.path.extension(command).isEmpty) {
+    extensions = platform.environment['PATHEXT'].split(pathSeparator);
+  }
+
+  List<String> candidates = <String>[];
+  if (command.contains(fileSystem.path.separator)) {
+    candidates = _getCandidatePaths(command, <String>[workingDirectory], extensions, fileSystem);
+  } else {
+    final List<String> searchPath = platform.environment['PATH'].split(pathSeparator);
+    candidates = _getCandidatePaths(command, searchPath, extensions, fileSystem);
+  }
+  for (final String path in candidates) {
+    final FileSystemEntityType type = fileSystem.typeSync(path);
+    if (type != FileSystemEntityType.notFound) {
+      return path;
+    }
+  }
+  logger.printTrace(
+    'Did not find executable for "$command", search path candidates: $candidates\n'
+    'Attempting original executable path.'
+  );
+  if (strict) {
+    return null;
+  }
+  return command;
```

Comment:
Maybe our logic is wrong and instead of throwing an argument error, we should just try to run it?

Reply:
Just trying to run it sounds reasonable to me, especially since we should be already handling the `ProcessException` everywhere we need to, but I'm less sure about the `ArgumentError`.